src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs

```diff
@@ -0,0 +1,181 @@
+        }
+
+        public static bool operator <=(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) <= 0;
+        }
+
+        public static SemanticVersion Parse(string semanticVersionString)
+        {
+            int majorSeparator = semanticVersionString.IndexOf(".");
+            if (majorSeparator == -1)
+            {
+                return null;
+            }
+
+            int major = 0;
+            if (!int.TryParse(semanticVersionString.Substring(0, majorSeparator), out major))
+            {
+                return null;
+            }
+
+            int minorStart = majorSeparator + 1;
+            int minorSeparator = semanticVersionString.IndexOf(".", minorStart);
+            if (minorSeparator == -1)
+            {
+                return null;
+            }
+
+            int minor = 0;
+            if (!int.TryParse(semanticVersionString.Substring(minorStart, minorSeparator - minorStart), out minor))
+            {
+                return null;
+            }
+
+            int patch = 0;
+            int patchStart = minorSeparator + 1;
+            int patchSeparator = semanticVersionString.FindFirstNotOf("0123456789", patchStart);
+            if (patchSeparator == -1)
+            {
+                if (!int.TryParse(semanticVersionString.Substring(patchStart), out patch))
+                {
+                    return null;
+                }
+
+                return new SemanticVersion(major, minor, patch);
+            }
+
+            if (!int.TryParse(semanticVersionString.Substring(patchStart, patchSeparator - patchStart), out patch))
+            {
+                return null;
+            }
+
+            int preStart = patchSeparator;
+            int preSeparator = semanticVersionString.IndexOf("+", preStart);
+            if (preSeparator == -1)
+            {
+                return new SemanticVersion(major, minor, patch, semanticVersionString.Substring(preStart));
+            }
+            else
+            {
+                int buildStart = preSeparator + 1;
+                return new SemanticVersion(
+                    major,
+                    minor,
+                    patch,
+                    semanticVersionString.Substring(preStart, preSeparator - preStart),
+                    semanticVersionString.Substring(buildStart));
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+            {
+                return false;
+            }
+
+            var other = obj as SemanticVersion;
+            if (other == null)
+            {
+                return false;
+            }
+
+            return this == other;
+        }
+
+        public override int GetHashCode()
+        {
+            return Major.GetHashCode() ^
+                   Minor.GetHashCode() ^
+                   Patch.GetHashCode() ^
+                   Pre.GetHashCode() ^
+                   Build.GetHashCode();
```

Comment:
This is not a very good hash code. I'd say `(Major, Minor, Patch, Pre).GetHashCode()` but we don't want the ValueTuple dependency here.

We don't need to hash semantic versions. I'd just remove all the overrides and overloads and keep just make Compare public. Better not to have dead internal code than to have to think about things like this that don't actually matter.